syntax = "proto3";

service Recommendations {
  rpc RecommendBook(RecommendationRequest) returns (RecommendationResponse) {};
  rpc GetAllBooksRecommend(GetAllBookRecommendationRequest)
      returns (GetAllBookRecommendationResponse) {};
  rpc CreateBook(CreateBookRequest) returns (CreateBookResponse) {};
  rpc UpdateBook(UpdateBookRequest) returns (UpdateBookResponse) {};
  rpc DeleteBook(DeleteBookRequest) returns (DeleteBookResponse) {};
}

message RecommendationRequest { int32 book_id = 1; }
message GetAllBookRecommendationRequest {}

message BookRecommendation {
  int32 id = 1;
  string title = 2;
}

message CreateBookRequest {
  string title = 2;
  string author = 3;
  string description = 4;
  string issue_date = 5;
  int32 price = 6;
  string category = 7;
}

message CreateBookResponse {
  int32 id = 1;
  string title = 2;
  string author = 3;
  string description = 4;
  string issue_date = 5;
  int32 price = 6;
  string category = 7;
}

message RecommendationResponse {
  int32 id = 1;
  string title = 2;
  string author = 3;
  string description = 4;
  string issue_date = 5;
  int32 price = 6;
  string category = 7;
}

message GetAllBookRecommendationResponse {
  repeated RecommendationResponse recommendations = 1;
}

message UpdateBookRequest {
  int32 id = 1;
  optional string title = 2;
  optional string author = 3;
  optional string description = 4;
  optional string issue_date = 5;
  optional int32 price = 6;
  optional string category = 7;
}

message UpdateBookResponse {
  int32 id = 1;
  string title = 2;
  string author = 3;
  string description = 4;
  string issue_date = 5;
  int32 price = 6;
  string category = 7;
}

message DeleteBookRequest { int32 id = 1; }
message DeleteBookResponse { string message = 1; }
